gtk4/src/subclass/accessible_text.rs

```diff
@@ -0,0 +1,533 @@
+
+use crate::accessible_text_range::AccessibleTextRange;
+use crate::subclass::prelude::*;
+use crate::{ffi, AccessibleText, AccessibleTextGranularity};
+use glib::object::Cast;
+use glib::translate::*;
+use glib::GString;
+
+pub trait AccessibleTextImpl: WidgetImpl {
+    #[doc(alias = "get_attributes")]
+    fn attributes(&self, offset: u32) -> Vec<(AccessibleTextRange, GString, GString)> {
+        self.parent_attributes(offset)
+    }
+
+    #[doc(alias = "get_caret_position")]
+    fn caret_position(&self) -> u32 {
+        self.parent_caret_position()
+    }
+
+    #[doc(alias = "get_contents")]
+    fn contents(&self, start: u32, end: u32) -> Option<glib::Bytes> {
+        self.parent_contents(start, end)
+    }
+
+    #[doc(alias = "get_contents_at")]
+    fn contents_at(
+        &self,
+        offset: u32,
+        granularity: crate::AccessibleTextGranularity,
+    ) -> Option<(u32, u32, glib::Bytes)> {
+        self.parent_contents_at(offset, granularity)
+    }
+
+    #[doc(alias = "get_default_attributes")]
+    fn default_attributes(&self) -> Vec<(GString, GString)> {
+        self.parent_default_attributes()
+    }
+
+    #[cfg(feature = "v4_16")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "v4_16")))]
+    #[doc(alias = "get_extents")]
+    fn extents(&self, start: u32, end: u32) -> Option<graphene::Rect> {
+        self.parent_extents(start, end)
+    }
+
+    #[cfg(feature = "v4_16")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "v4_16")))]
+    #[doc(alias = "get_offset")]
+    fn offset(&self, point: &graphene::Point) -> Option<u32> {
+        self.parent_offset(point)
+    }
+
+    #[doc(alias = "get_selection")]
+    fn selection(&self) -> Vec<AccessibleTextRange> {
+        self.parent_selection()
+    }
+}
+
+mod sealed {
+    pub trait Sealed {}
+    impl<T: super::AccessibleTextImplExt> Sealed for T {}
+}
+
+pub trait AccessibleTextImplExt: sealed::Sealed + ObjectSubclass {
+    fn parent_attributes(&self, offset: u32) -> Vec<(AccessibleTextRange, GString, GString)> {
+        unsafe {
+            let type_data = Self::type_data();
+            let parent_iface = type_data.as_ref().parent_interface::<AccessibleText>()
+                as *const ffi::GtkAccessibleTextInterface;
+
+            let func = (*parent_iface)
+                .get_attributes
+                .expect("no parent \"get_attributes\" implementation");
+
+            let mut n_ranges = MaybeUninit::uninit();
+            let mut ranges = std::ptr::null_mut();
+            let mut attribute_names = std::ptr::null_mut();
+            let mut attribute_values = std::ptr::null_mut();
+
+            let is_set: bool = from_glib(func(
+                self.obj()
+                    .unsafe_cast_ref::<AccessibleText>()
+                    .to_glib_none()
+                    .0,
+                offset,
+                n_ranges.as_mut_ptr(),
+                &mut ranges,
+                &mut attribute_names,
+                &mut attribute_values,
```

Comment:
I don't think there are necessarily `n_ranges` attribute names/values. Both are just `NULL`-terminated string arrays. I'd assume both have the same length but otherwise they should be independent from the ranges